domain/storage/service/storage_test.go

```diff
@@ -0,0 +1,281 @@
+}
+
+type filesystemImporter struct {
+	*MockFilesystemSource
+	*MockFilesystemImporter
+}
+
+func (s *storageSuite) setupMocks(c *gc.C) *gomock.Controller {
+	ctrl := gomock.NewController(c)
+
+	s.state = NewMockState(ctrl)
+	s.volumeSource = NewMockVolumeSource(ctrl)
+	s.volumeImporter = NewMockVolumeImporter(ctrl)
+	s.filesystemSource = NewMockFilesystemSource(ctrl)
+	s.filesystemImporter = NewMockFilesystemImporter(ctrl)
+	s.provider = NewMockProvider(ctrl)
+
+	registry := NewMockProviderRegistry(ctrl)
+	registry.EXPECT().StorageProvider(storage.ProviderType("ebs")).Return(s.provider, nil).AnyTimes()
+	registry.EXPECT().StorageProvider(storage.ProviderType("elastic")).Return(s.provider, nil).AnyTimes()
+
+	s.registry = storage.ChainedProviderRegistry{registry, provider.CommonStorageProviders()}
+
+	return ctrl
+}
+
+func (s *storageSuite) service(c *gc.C) *Service {
+	return NewService(s.state, loggertesting.WrapCheckLog(c), modelStorageRegistryGetter(func() storage.ProviderRegistry {
+		return s.registry
+	}))
+}
+
+func (s *storageSuite) TestImportFilesystemValidate(c *gc.C) {
+	defer s.setupMocks(c).Finish()
+
+	_, err := s.service(c).ImportFilesystem(context.Background(), ImportStorageParams{
+		Kind:        storage.StorageKindFilesystem,
+		Pool:        "elastic",
+		ProviderId:  "provider-id",
+		StorageName: "0",
+	})
+	c.Assert(err, jc.ErrorIs, corestorage.InvalidStorageName)
+
+	_, err = s.service(c).ImportFilesystem(context.Background(), ImportStorageParams{
+		Kind:        storage.StorageKindFilesystem,
+		Pool:        "0",
+		ProviderId:  "provider-id",
+		StorageName: "pgdata",
+	})
+	c.Assert(err, jc.ErrorIs, storageerrors.InvalidPoolNameError)
+
+	_, err = s.service(c).ImportFilesystem(context.Background(), ImportStorageParams{
+		Kind:        storage.StorageKindBlock,
+		Pool:        "elastic",
+		ProviderId:  "provider-id",
+		StorageName: "pgdata",
+	})
+	c.Assert(err, jc.ErrorIs, errors.NotSupported)
+}
+
+func (s *storageSuite) TestImportFilesystem(c *gc.C) {
+	defer s.setupMocks(c).Finish()
+
+	s.provider.EXPECT().Supports(storage.StorageKindFilesystem).Return(true)
+	cfg, err := storage.NewConfig("elastic", "elastic", nil)
+	c.Assert(err, jc.ErrorIsNil)
+	s.provider.EXPECT().FilesystemSource(cfg).Return(filesystemImporter{
+		MockFilesystemSource:   s.filesystemSource,
+		MockFilesystemImporter: s.filesystemImporter,
+	}, nil)
+
+	s.state.EXPECT().GetStoragePoolByName(gomock.Any(), "elastic").Return(domainstorage.StoragePoolDetails{}, storageerrors.PoolNotFoundError)
+	s.state.EXPECT().GetModelDetails().Return(domainstorage.ModelDetails{
+		ModelUUID:      coretesting.ModelTag.Id(),
+		ControllerUUID: coretesting.ControllerTag.Id(),
+	}, nil)
+	s.filesystemImporter.EXPECT().ImportFilesystem(gomock.Any(), "provider-id", map[string]string{
+		"juju-model-uuid":      coretesting.ModelTag.Id(),
+		"juju-controller-uuid": coretesting.ControllerTag.Id(),
+	}).Return(storage.FilesystemInfo{
+		FilesystemId: "provider-id",
```

Comment:
Would it be adventurous to use a different id here in the tests from that of the provider so when a failure happens it is possible to tell which value is which in test output?

Reply:
sure